rust/actyx/swarm/harness/src/lib.rs

```diff
@@ -299,42 +300,74 @@ pub async fn fully_meshed<E>(sim: &mut Netsim<Command, E>, timeout: Duration) ->
-    futures::pin_mut!(deadline);
     let peers = sim.machines().iter().map(|m| m.peer_id()).collect::<Vec<_>>();
-    for (idx, machine) in sim.machines_mut().iter_mut().enumerate() {
-        let mut peers = peers
-            .iter()
-            .filter(|p| **p != peers[idx])
-            .copied()
-            .collect::<BTreeSet<_>>();
-        while !peers.is_empty() {
-            let res = {
-                let f = machine.select(|ev| m!(ev.borrow(), Event::Connected(p) => *p));
-                futures::pin_mut!(f);
-                match select(deadline.as_mut(), f).await {
-                    Either::Left(_) => Either::Left(()),
-                    Either::Right(r) => Either::Right(r),
-                }
-            };
-            match res {
-                Either::Left(_) => {
-                    for e in machine.drain() {
-                        tracing::error!("got event {}", e);
-                    }
-                    bail!(
-                        "fully_meshed timed out after {:.1}sec ({}, {:?})",
-                        timeout.as_secs_f64(),
-                        idx,
-                        peers
-                    )
-                }
-                Either::Right((None, _)) => bail!("got not peer"),
-                Either::Right((Some(p), _)) => {
-                    peers.remove(&p);
+
+    let machines_promises = sim
+        .machines_mut()
+        .iter_mut()
+        .enumerate()
+        .map(|(idx, machine)| -> BoxFuture<Result<()>> {
+            let mut peers = peers
+                .iter()
+                .filter(|p| **p != peers[idx])
+                .copied()
+                .collect::<BTreeSet<_>>();
+            async move {
+                let deadline = task::sleep(timeout);
+                futures::pin_mut!(deadline);
+                while !peers.is_empty() {
+                    let res = {
+                        let f = async {
+                            let x = machine.select(|ev| m!(ev.borrow(), Event::Connected(p) => *p));
+                            futures::pin_mut!(x);
+                            loop {
+                                let poll = futures::poll!(&mut x);
+                                match poll {
+                                    task::Poll::Ready(x) => return x,
+                                    task::Poll::Pending => {
+                                        tokio::task::yield_now().await;
+                                        // tokio::time::sleep(Duration::from_millis(20)).await;
+                                    }
+                                };
+                            }
+                        };
+
+                        futures::pin_mut!(f);
+                        match select(deadline.as_mut(), f).await {
+                            Either::Left(_) => Either::Left(()),
+                            Either::Right(r) => Either::Right(r),
+                        }
+                    };
```

Comment:
This should really just be
```rust
let f = machine.select(...);
futures::pin_mut!(f);
let res = select(deadline.as_mut(), f).await;
```
If changing it to the code above changes the behaviour, then something really weird must be broken in strange ways. And the breakage is most likely in `machine.select`.

Reply:
This got me a compile error due to multiple mutable references, But I'll remove the inner polling future